pkg/app/piped/cloudprovider/lambda/client.go

```diff
@@ -405,6 +409,100 @@ func (c *client) UpdateTrafficConfig(ctx context.Context, fm FunctionManifest, r
 	return nil
 }
 
+func (c *client) updateTagsConfig(ctx context.Context, fm FunctionManifest) error {
+	getFuncInput := &lambda.GetFunctionInput{
+		FunctionName: aws.String(fm.Spec.Name),
+	}
+	output, err := c.client.GetFunctionWithContext(ctx, getFuncInput)
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			case lambda.ErrCodeInvalidParameterValueException:
+				return fmt.Errorf("invalid parameter given: %w", err)
+			case lambda.ErrCodeServiceException:
+				return fmt.Errorf("aws lambda service encountered an internal error: %w", err)
+			case lambda.ErrCodeTooManyRequestsException:
+				return fmt.Errorf("request throughput limit was exceeded: %w", err)
+			case lambda.ErrCodeResourceNotFoundException:
+				return fmt.Errorf("resource not found: %w", err)
+			}
+		}
+		return fmt.Errorf("unknown error occurred on list tags of Lambda function %s: %w", fm.Spec.Name, err)
+	}
+
+	functionArn := aws.StringValue(output.Configuration.FunctionArn)
+	currentTags := aws.StringValueMap(output.Tags)
+	// Skip if there are no changes on tags.
+	if reflect.DeepEqual(currentTags, fm.Spec.Tags) {
+		return nil
+	}
+
+	// Skip untag resource if remote resource has no tags.
+	if len(output.Tags) == 0 {
+		goto Tags
+	}
+
+	{ // Untag previous tags from the current resource.
```

Comment:
As I see the current approach is removing all existing tags and adding the defined ones, right?
This way makes a small amount of time when no tags are assigned to the function.
I think that is no big deal in most cases but maybe sometimes that is not good.
So maybe we should change to
- adding newly added tags
- removing no-longer defined tags

Reply:
thanks for your comment 🙇 
There is a case leaked from your suggestion, that is the same tags key but the value of those tags are changed. 🤔 
In that case, we have to remove changed tags and adding newly defined values for those tags (was API does not provide ways to update existing tags). For the cost of compare which tags are updated, defined, removed, etc, I think just removing all existing tags and adding all as defined in `function.yaml` is simpler and can keep defined tags strict with the remote newly deploy function. How do you think about it? 👀

Reply:
>  I think just removing all existing tags and adding all as defined in function.yaml is simpler
 
Yes, I agree about this.

I am figuring out the usage of these tags. As the docs from AWS
> Tags are freeform key-value pairs that are supported across AWS services for use in filtering resources and adding detail to billing reports.

Besides, I think they can be used for monitoring and alerting.

Actually, I still have the feeling that removing all existing ones then re-adding the defined ones is not good.
Maybe it is not what the users expected. They expect the tags should be updated smoothly, without any no-tag moment.

It is just a small thing but I think it is worth doing to avoid unexpected behavior.

The tags are just a list of key-value, so we can compare them to generate added-ones, changed-ones, removed-ones, and then
- adding newly added ones
- updating changed ones
- deleting removed-ones

How do you think?

Reply:
> They expect the tags should be updated smoothly, without any no-tag moment.

I concern the same thing too 👍 The updating sequence (in my guess) should be
- adding newly added ones
- remove the changed ones
- add the changed ones
- deleting removed-ones
since we do not have the tags updating API 😢 Still have a bit of leak since alerting based on `remove the changed ones`, but at least it avoids the `no tags` moment 👍 
I will go with your suggestion 🙏

Reply:
> since we do not have the tags updating API

Lol. Right.

> I will go with your suggestion 🙏

Thank you.

Reply:
addressed on https://github.com/pipe-cd/pipe/pull/1596/commits/a9d93fa35b7bba4b92e38d508eec8602d1b6dabc 🙏